docs/introduction/how-to-work-with-money.md

```diff
@@ -0,0 +1,388 @@
+The only difference is that it outputs the text *"Free"* when zero amount of money is provided.
+
+### priceTextWithCurrencyByCurrencyIdAndLocale
+
+Filter `priceTextWithCurrencyByCurrencyIdAndLocale` can be used to format the amount of money in any currency and locale.
+It outputs the text *"Free"* when zero amount of money is provided.
+
+The *currency ID* (`int`) and the *locale* (`string`) must be provided as parameters.
+
+### priceWithCurrency
+
+Filter `priceWithCurrency` can be used to format the amount of money in any currency.
+It uses the current locale (the locale of current domain or administration) to format it.
+
+The *currency* (`Currency` entity) must be provided as a parameter.
+
+### priceWithCurrencyAdmin
+
+It works similarly to `priceWithCurrency`, but it uses the default administration currency.
+
+It has no parameters.
+
+### priceWithCurrencyByCurrencyId
+
+It works similarly to `priceWithCurrency`, but it uses currency ID instead of the `Currency` entity.
+
+The *currency ID* (`int`) must be provided as a parameter.
+
+### priceWithCurrencyByDomainId
+
+It works similarly to `priceWithCurrency`, but it uses the basic frontend currency of the provided domain.
+
+The *domain ID* (`int`) must be provided as a parameter.
+
+## Money in Javascript
+
+`Money` implements the `JsonSerializable` interface, which means it can be serialized into JSON format with `json_encode` into an object with the following structure:
+
+```json
+{
+  "amount": "1234.5670"
+}
+```
+
+This format can be readily used in Javascript when JSON is rendered in Twig:
+
+```twig
+{# the "money" variable contains Money::create('1234.5670') #}
+
+<div id="my-money" data-money="{{ money|json_encode }}"/>
+```
+
+```js
+(function ($) {
+
+    $(document).ready(function () {
+        var money = $('#my-money').data('money');
+
+        // Displays "1234.5670"
+        alert(money.amount);
+    });
+
+})(jQuery);
+```
+
+See the native Javascript method [`.toFixed()`](https://www.w3schools.com/jsref/jsref_tofixed.asp) and the function [`parseFloat()`](https://www.w3schools.com/jsref/jsref_parsefloat.asp) to see how to convert between decimal strings and numbers.
+
+## Money in Doctrine
+
+In Shopsys Framework, there is a custom Doctrine type `money` which can be used in similar fashion as `decimal` type.
+The entity property value will be automatically hydrated to an instance of `Money` if it's configured to use the type:
+
+```php
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * @ORM\Table(name="my_entities")
+ * @ORM\Entity
+ */
+class MyEntity
+{
+    /**
+     * @var \Shopsys\FrameworkBundle\Component\Money\Money|null
+     *
+     * @ORM\Column(type="money", precision=20, scale=6, nullable=true)
+     */
+    protected $price;
+
+    // ...
+}
+```
+
+When you want to use a `Money` instance as a parameter in DQL, use the `getAmount` method in your [repository class](/docs/introduction/basics-about-model-architecture.md#repository):
```

Comment:
I do not want to be a nitpick, however, I would expect this to work automatically provided that `ProductCalculatedPrice::$priceWithVat` is `Money` type :slightly_smiling_face:

Reply:
We've expected that too, that's why I've included it in the docs. There are possibilities how to do that in Doctrine, but you either have to implement `__toString` in your value object (which we wanted to avoid), or you have to call `
Type::getType('money')->convertToDatabaseValue($money, $platform)`, but that requires having an instance of `\Doctrine\DBAL\Platforms\AbstractPlatform` and is rather complicated...

Setting it as a parameter and getting it out of aggregate functions (such as `MAX()` or `SUM()`) are two issues with Doctrine we run into. Converting it to/from `Money` in a repository is a decent solution though.

Reply:
Maybe we should document (or at least mention) working with aggregate functions as well...

Reply:
I've tried to explain it in f446f4082